pkg/engine/lifeycletest/resource_reference_test.go

```diff
@@ -0,0 +1,221 @@
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			v := &deploytest.Provider{
+				CreateF: func(urn resource.URN, news resource.PropertyMap,
+					timeout float64, preview bool) (resource.ID, resource.PropertyMap, resource.Status, error) {
+
+					id := "created-id"
+					if preview {
+						id = ""
+					}
+
+					state := resource.PropertyMap{}
+					if urn.Name() == "resC" {
+						state = resource.PropertyMap{
+							"resA": resource.MakeResourceReference(urnA, "", false, ""),
+							"resB": resource.MakeResourceReference(urnB, idB, true, ""),
+						}
+					}
+
+					return resource.ID(id), state, resource.StatusOK, nil
+				},
+				ReadF: func(urn resource.URN, id resource.ID,
+					inputs, state resource.PropertyMap) (plugin.ReadResult, resource.Status, error) {
+					return plugin.ReadResult{Inputs: inputs, Outputs: state}, resource.StatusOK, nil
+				},
+			}
+			return v, nil
+		}),
+	}
+
+	opts := deploytest.ResourceOptions{DisableResourceReferences: true}
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		var err error
+		urnA, _, _, err = monitor.RegisterResource("component", "resA", false, opts)
+		assert.NoError(t, err)
+
+		err = monitor.RegisterResourceOutputs(urnA, resource.PropertyMap{})
+		assert.NoError(t, err)
+
+		urnB, idB, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, opts)
+		assert.NoError(t, err)
+
+		_, _, props, err := monitor.RegisterResource("pkgA:m:typA", "resC", true, opts)
+		assert.NoError(t, err)
+
+		assert.True(t, props.DeepEquals(resource.PropertyMap{
+			"resA": resource.NewStringProperty(string(urnA)),
+			"resB": resource.NewStringProperty(string(idB)),
+		}))
+		return nil
+	})
+	host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+
+	p := &TestPlan{
+		Options: UpdateOptions{Host: host},
+		Steps:   MakeBasicLifecycleSteps(t, 4),
+	}
+	p.Run(t, nil)
+}
+
+// TestResourceReferences_DownlevelEngine tests an SDK that supports resource references communicating with an engine
+// that does not.
+func TestResourceReferences_DownlevelEngine(t *testing.T) {
+	var urnA resource.URN
+	var urnB resource.URN
+	var idB resource.ID
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			v := &deploytest.Provider{
+				CreateF: func(urn resource.URN, news resource.PropertyMap,
+					timeout float64, preview bool) (resource.ID, resource.PropertyMap, resource.Status, error) {
+
+					id := "created-id"
+					if preview {
+						id = ""
+					}
+
+					// If we have resource references here, the engine has not properly disabled them.
+					if urn.Name() == "resC" {
+						assert.Equal(t, resource.NewStringProperty(string(urnA)), news["resA"])
+						assert.Equal(t, resource.NewStringProperty(string(idB)), news["resB"])
```

Comment:
I don't think I quite understand this. The resource can be referenced by either the id or the urn? I would have expected this to always be via the urn? Sorry if this is simple, but I'm just not familiar with this part of the code very much.

Reply:
> // TestResourceReferences_DownlevelEngine tests an SDK that supports resource references communicating with an engine
 // that does not.

This test simulates an SDK that supports resource references communicating with an engine that does not support resource references. In that case, the marshaller should transform a resource reference into either a plain URN (for references to component resources) or a plan ID (for references to custom resources). These checks validate that the transformation occurred as expected.